include/deal.II/fe/fe_poly.templates.h

```diff
@@ -251,19 +256,26 @@ fill_fe_values (const Mapping<dim,spacedim>                                  &ma
 
       if (flags & update_hessians && cell_similarity != CellSimilarity::translation)
         {
-          // compute the hessians in the unit cell (accounting for the Jacobian gradiant)
-          for (unsigned int i=0; i<quadrature.size(); ++i)
-            {
-              fe_data.untransformed_shape_hessians[i] = fe_data.shape_hessians[k][i];
-            }
-
-          correct_untransformed_hessians (fe_data.untransformed_shape_hessians,
-                                          mapping_data, output_data, quadrature.size(), k);
-
-          mapping.transform (fe_data.untransformed_shape_hessians,
+          mapping.transform (fe_data.shape_hessians[k],
                              mapping_covariant_gradient,
                              mapping_internal,
                              output_data.shape_hessians[k]);
+
+          for (unsigned int i=0; i<quadrature.size(); ++i)
+            for (unsigned int m=0; m<spacedim; ++m)
+              output_data.shape_hessians[k][i] -=
+                mapping_data.jacobian_pushed_forward_grads[i][m]
+                * output_data.shape_gradients[k][i][m];
```

Comment:
I think this could probably be written in more readable form as
```
  for (i=....)
    for (j=...)
      output.shape_hessians[k][i] 
         -= mapping_data.jacobian_pushed_forward_grads[i] * output_data.shape_gradients[k][i];
```
In general (as a consideration for future patches, not as a request to make the current patch even more complex), I think many of these deeply nested loops could be written in a more readable way by using `operator*` or one of the `contract(...)` variants on the underlying tensors.